test/e2e/ingressdns.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"sort"
+
+	apiv1 "k8s.io/api/core/v1"
+	extv1b1 "k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	dnsv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/multiclusterdns/v1alpha1"
+	fedclientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	"github.com/kubernetes-sigs/federation-v2/test/common"
+	"github.com/kubernetes-sigs/federation-v2/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+)
+
+var _ = Describe("MultiClusterIngressDNS", func() {
+	f := framework.NewFederationFramework("multicluster-ingress-dns")
+	tl := framework.NewE2ELogger()
+
+	const userAgent = "test-ingress-dns"
+	const baseName = "test-ingress-dns-"
+
+	var fedClient fedclientset.Interface
+	var clusterRegionZones map[string]fedv1a1.FederatedClusterStatus
+	var objectGetter func(namespace, name string) (pkgruntime.Object, error)
+
+	BeforeEach(func() {
+		fedClient = f.FedClient(userAgent)
+		// TODO(marun) How to discover the namespace in an unmanaged scenario?
+		federatedClusters, err := fedClient.CoreV1alpha1().FederatedClusters(f.FederationSystemNamespace()).List(metav1.ListOptions{})
+		framework.ExpectNoError(err, "Error listing federated clusters")
+		clusterRegionZones = make(map[string]fedv1a1.FederatedClusterStatus)
+		for _, cluster := range federatedClusters.Items {
+			clusterRegionZones[cluster.Name] = fedv1a1.FederatedClusterStatus{
+				Region: cluster.Status.Region,
+				Zone:   cluster.Status.Zone,
+			}
+		}
+		f.SetUpDNSControllerFixture()
+
+		objectGetter = func(namespace, name string) (pkgruntime.Object, error) {
```

Comment:
> Why does this need to be defined in a BeforeEach closure? It doesn't appear to have any dependency on closure variables.

it is a common variable used in multiple test cases. makes sense to define it once in `BeforeEach`.

> (No action required) Maybe it would be simpler to define a variable of type MultiClusterIngressDNSRecordInterface?

Updated.

Reply:
If a variable varies across test cases it belongs in BeforeEach.  If it doesn't change, I think it's preferable to include it outside of that section (since it only needs to execute once rather than before every test).

Reply:
ya you are right. I somehow misunderstood it. updated now